### ISSUE TITLE
Character escape regression with v0.10.10

### ISSUE THREAD

**Issue by n7s** [silnrsi/fontproof#17]
This might well be very obvious but anyway here goes: 

A small regression has been flagged up by our CI:

this is confirmed to work with sile v0.10.9:
escaping % with a backslash in punctuation patterns where we want the special character to appear as is.

but it does not work anymore with sile v0.10.10:

Error detected:
fontproof.lua:259: invalid use of '%' in replacement string

What has overriden the escaping of special characters? 
A basic syntax issue we have overlooked? 
Any pointers?

**Comment by alerque** [comment]
This almost certainly has to do with the change in https://github.com/sile-typesetter/sile/pull/997. The TL;DR version is that special character handling (like `\%`) got upgraded from a combination of 3 hacks (1 to ignore them in the parser, 2 being to override what was allowed in command names, 3 being to register dummy pass-through commands to make them non-functional) to a first class part of the initial syntax parsing for `sil` language input.

In our test suite the only thing that changed to accommodate this was some minor metric shifts for fonts that handled kerning differently for a single shaping run on things like `{h` than for two consecutive runs on `{` and `h`). I considered this a win, not a loss.

Can you be more specific here about a test case that fails? It sounds like you're passing a string somewhere than is being interpreted as a Lua pattern, but without an MWE it's pretty hard to tell.

**Comment by alerque** [comment]
The registered command `\pattern` is taking *content* and passing it to Lua's gsub as a pattern. The `%` character has special meaning in both the `*.sil` input format (where it is a comment character) and Lua's patterns (where it is an escape character). It has to be escaped for use in both. Given that **where** SILE unescapes percent signs has changed (each command now gets clean content with all removal of escape symbols complete, previously it might have gotten passed content that had sub-commands that called a percent symbol output command) I'm assuming your strings are now being passed in a more raw form that you expect.

A full working example would help me track down how to fix this for you.

**Comment by n7s** [comment]
Thanks so much for digging into this.

One example is https://github.com/silnrsi/font-shimenkan/blob/9b502774de7c7f61895d70fad87eb3dd402069d8/tests/MiaoTestPunctNum.sil#L13

**Comment by n7s** [comment]
I suspect better wrapping of the string would help, but I couldn't figure it out.

**Comment by alerque** [comment]
Try changing the `%` to `%%`. That's the Lua escape to get a literal percent sign.

Honestly I would have written this differently and added an argument to `\pattern` and passed that stuff as a string instead of re-using the input content for something that isn't valid as or processed as SILE input.

**Comment by n7s** [comment]
`%%` gives a parse error (tried that earlier).

**Comment by DavidLRowe** [comment]
Does "\037" work to get "%"? (decimal 37 = hex 25, for U+0025 "%")

**Comment by alerque** [comment]
> `%%` gives a parse error (tried that earlier). 

How about `\%\%`? One for the SILE input, one for Lua's gsub pattern. Sorry I'm running around on mobile and not in a position to try for a while.

**Comment by devosb** [comment]
I tested `\%\%` and it behaves like two separate instances of `\%`. So it does not solve the issue, neither did any of the other suggestions. With running _sile_ from the SILE PPA on Ubuntu Focal I saw the same issue that @n7s did. Running _sile_ compiled from source (both the latest and the v0.10.10 tag) with `luarocks` to install the needed dependencies did not show the error, but the PDF output was not correct for `%`.
